File: bus/trigger.go
package bus

import (
	"math"
	"strings"
)

type Trigger struct {
	Topic     string
	handler   Handler
	topicList []string
	Id        uint32
	wildCard  bool
}

func NewTrigger(topic string, handler Handler) *Trigger {
	trigger := &Trigger{
		Topic:     topic,
		handler:   handler,
		topicList: strings.Split(topic, "."),
		wildCard:  false,
	}
	for _, t := range trigger.topicList {
		if t == "*" {
			trigger.wildCard = true
		}
	}

	return trigger

}

func (t *Trigger) Compare(topicList []string) bool {
	triggerLength := len(t.topicList)
	topicListLength := len(topicList)
	if !t.wildCard && triggerLength != topicListLength {
		return false
	}
	maxLoop := uint(math.Min(float64(triggerLength), float64(topicListLength)))
	var i uint = 0
	for ;i < maxLoop;i++  {
		if t.topicList[i] == "*" {
			break
		}
		if t.topicList[i] != topicList[i] {
			return false
		}
	} 

	return true
}

File: src/main/java/ovh/angrysoft/homedaemon/bus/Trigger.java
package ovh.angrysoft.homedaemon.bus;

public class Trigger {
    private Handler handler;
    private String[] topicList;
    private boolean wildCard;
    private boolean oneShot;
    private String id = "";

    public Trigger(String topic, Handler handler) {
        this(topic, handler, false);
    }

    public Trigger(String topic, Handler handler, boolean oneShot) {
        this.oneShot = oneShot;
        topicList = topic.split("\\.");
        this.handler = handler;
        this.wildCard = false;
        for (String subTopic : topicList) {
            if (subTopic.equals("*"))
                this.wildCard = true;
        }
    }

    // TODO seperate class topic topic.equals
    public boolean compareTopic(String[] topicList) {
        if (!this.wildCard && topicList.length != this.topicList.length)
            return false;
        int maxLoop = Math.min(this.topicList.length, topicList.length);
        for (int i = 0; i < maxLoop; i++) {
            if (this.topicList[i].equals("*"))
                break;
            if (!this.topicList[i].equals(topicList[i]))
                return false;
        }
        return true;
    }

    public void call(Event event) {
        handler.call(event);
    }

    public boolean isOneShot() {
        return oneShot;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }
}
